pkg/sql/contention/txnidcache/fifo_cache.go

```diff
@@ -0,0 +1,234 @@
+// Copyright 2022 The Cockroach Authors.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package txnidcache
+
+import (
+	"sync"
+
+	"github.com/cockroachdb/cockroach/pkg/roachpb"
+	"github.com/cockroachdb/cockroach/pkg/sql/contention/contentionutils"
+	"github.com/cockroachdb/cockroach/pkg/util/syncutil"
+	"github.com/cockroachdb/cockroach/pkg/util/uuid"
+	"github.com/cockroachdb/errors"
+)
+
+type fifoCache struct {
+	pool     *sync.Pool
+	capacity contentionutils.CapacityLimiter
+
+	mu struct {
+		syncutil.RWMutex
+
+		data map[uuid.UUID]roachpb.TransactionFingerprintID
+
+		eviction evictionList
+	}
+}
+
+type evictionNode struct {
+	block *messageBlock
+	next  *evictionNode
+}
+
+// evictionList is a singly-linked list of messageBlocks. The list is used to
+// implement FIFO eviction.
+type evictionList struct {
+	head *evictionNode
+	tail *evictionNode
+
+	// tailBlockIdx is an index pointing into the next empty slot in messageBlock
+	// stored in the tail pointer.
+	tailBlockIdx int
+
+	scratchZeroBlock messageBlock
+}
+
+func newFIFOCache(pool *sync.Pool, capacity contentionutils.CapacityLimiter) *fifoCache {
+	c := &fifoCache{
+		pool:     pool,
+		capacity: capacity,
+	}
+
+	c.mu.data = make(map[uuid.UUID]roachpb.TransactionFingerprintID)
+	return c
+}
+
+// Add insert a messageBlock into the fifoCache.
+//
+// N.B. After Add() returns, the input messageBlock should not be reused. The
+// fifoCache sometimes directly hijack the input messageBlock to form the
+// eviction list and often even directly overwrite its content to minimize
+// memory allocation.
+func (c *fifoCache) Add(block *messageBlock) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	blockSize := 0
+	for i := range block {
+		if !block[i].valid() {
+			break
+		}
+
+		c.mu.data[block[i].TxnID] = block[i].TxnFingerprintID
+		blockSize++
+	}
+
+	if blockHijacked := c.mu.eviction.insertBlock(block, blockSize); !blockHijacked {
```

Comment:
some of this hijacking feels too clever to me. I feel like just embedding the `messageBlock` as a value in eviction list and pooling that will be simpler from a data abstraction perspective layer. You can do it as a single for loop over the block. I don't feel that strongly though.

Reply:
Done. Switched to using `pool` for node allocation.